### ISSUE TITLE
Major Android Update Discussion: Introducing Offline Support and Resolving CORS Issues

### ISSUE THREAD

**Issue by jiongxuan** [johannesjo/super-productivity#3527]
### **Problem:**

- **Lack of Offline Support:** The existing Android application requires an active internet connection to function, preventing users from accessing core features when offline.

- **CORS (Cross-Origin Resource Sharing) Issues:** The app encounters CORS-related problems, limiting its ability to interact securely with local or self-hosted resources.

- **Security Concerns:** There are significant security vulnerabilities related to network hijacking, which could lead to data breaches or unauthorized access if the app relies solely on external servers without proper safeguards.

---

### **Objective:**

- **Enable Offline Functionality:** Allow users to utilize the Android app's core features without needing an internet connection, enhancing accessibility and reliability.

- **Resolve CORS Issues:** Configure WebView appropriately to eliminate CORS-related problems, ensuring secure and flexible interactions with various resources.

- **Enhance Security:** Implement measures to prevent network hijacking and other security threats, ensuring that data interactions remain secure even when offline.

---

### **User Experience:**

- **Enhanced Accessibility:** Users can now access and use the app's functionalities even without an internet connection, making the app more reliable and user-friendly.

- **Seamless Experience for Existing Users:** Existing users upgrading from older versions will continue using the app without any disruptions, maintaining data integrity and consistent functionality.

- **Offline Mode for New Installations:** New installations will have offline capabilities enabled by default, allowing users to work without constant internet access.

- **Future Migration Plans:** While the new offline solution is available for fresh installations, plans are in place to develop migration strategies to transition existing users to the new offline mode without affecting their current usage.

---

### **Technical Changes:**

- **Introduced `LaunchDecider` Class:**
  - **Purpose:** Manages the app's launch behavior based on whether the user performed a fresh installation or upgraded from a previous version.
  - **Functionality:** Utilizes `SharedPreferences`, `PackageInfo`, and file existence checks to determine the appropriate launch mode (`MODE_ONLINE` for existing users and `MODE_OFFLINE` for new installations).
  - **Implementation:** Ensures that users are directed to the correct activity (`FullscreenActivity` or `CapacitorMainActivity`) based on their installation status.

- **Extracted `WebViewRequestHandler`:**
  - **Purpose:** Centralizes the handling of WebView requests, including URL loading and request interception, to avoid code duplication across activities.
  - **Functionality:** Manages both `shouldOverrideUrlLoading` and `shouldInterceptRequest` methods, ensuring consistent behavior and streamlined request processing.

- **Introduced Capacitor:**
  - **Reasoning:** Compared to Cordova, Capacitor offers better performance, a more modern architecture, and easier integration with native platforms. Capacitor provides a more robust bridge for web-native interactions, enhancing the app's capabilities and maintainability.

- **Added `CapacitorMainActivity`:**
  - **Purpose:** Isolates the new Capacitor-based implementation from the existing `FullscreenActivity`, preventing mutual interference and ensuring stable operation for both new and existing users.

- **Updated `AndroidManifest.xml`:**
  - **Configuration:** Set `FullscreenActivity` as the default launcher activity and declared `CapacitorMainActivity` appropriately without a launcher intent filter.
  - **Intent Handling:** Ensured that `FullscreenActivity` manages the transition to `CapacitorMainActivity` based on the launch decision logic.

- **Refactored `FullscreenActivity`:**
  - **Integration:** Incorporated `LaunchDecider` to determine the appropriate activity to launch during `onCreate`.
  - **Stability Fix:** Added checks to prevent `wvContainer` from causing crashes by verifying its initialization before attempting to remove views in `onDestroy`.

---

### **Progress:**

- **Development Progress:** 80% completed. The primary functionalities for offline support and CORS resolution have been implemented. Remaining tasks include minor adjustments and thorough testing to ensure stability. See: 

    - https://github.com/jiongxuan/super-productivity-android/tree/feat/platform-android-offline
    - https://github.com/jiongxuan/super-productivity/tree/feat/platform-android-offline

- **Upcoming Plans:** Aiming to finalize development and resolve outstanding issues by October, with plans to release the update for user availability.

---

### **Related Issues:**

- [App fails to start without internet connection #3](https://github.com/super-productivity/super-productivity-android/issues/3)
- [Serve web app locally #4](https://github.com/super-productivity/super-productivity-android/issues/4)
- [CORS headers should not be required #16](https://github.com/super-productivity/super-productivity-android/issues/16)

---

### **Conclusion:**

I have made significant progress in addressing longstanding issues related to offline support and CORS in the Android application. These enhancements aim to improve user experience by enabling offline functionality, resolving security-related CORS problems, and ensuring a seamless transition for existing users. 

Welcome any questions or feedback from the community to ensure the update meets user needs and expectations.

**Comment by ElpyGh** [comment]
For your information, I made the following test:,for Android version 11.1.2 L1A
1.. restart the mobile (Android 13)
2. Disable wifi in the mobile
3. Start SuperProductivity app: the app doesn't start. Insead it appears a screen wit the message:
"Web page not available"
The web page at https://app.super-productivity.com could not be loaded because:
net:ERR_INTERNET_DISCONNECTED
I also support the idea that the SP Android app, to be consistent with the privacy claims, should be able to run standalone without the need to connect to internet.

**Comment by jiongxuan** [comment]
> Seamless Experience for Existing Users: Existing users upgrading from older versions will continue using the app without any disruptions, maintaining data integrity and consistent functionality.



> For your information, I made the following test:,for Android version 11.1.2 L1A 1.. restart the mobile (Android 13) 2. Disable wifi in the mobile 3. Start SuperProductivity app: the app doesn't start. Insead it appears a screen wit the message: "Web page not available" The web page at https://app.super-productivity.com could not be loaded because: net:ERR_INTERNET_DISCONNECTED I also support the idea that the SP Android app, to be consistent with the privacy claims, should be able to run standalone without the need to connect to internet.

@ElpyGh This behavior occurs because the app needs to be reinstalled to switch to a fully offline mode. This is a compatibility measure to avoid potential issues with data integrity when users suddenly switch to a new environment. In the future, I plan to implement a migration mode that will allow seamless runtime switching without requiring a reinstall.

**Comment by ElpyGh** [comment]
Ok, so because I use the Android app that doesn't provide the export/import feature, I should synchronize with a laptop, then remove the Android app, reinstall it and then synchronize again from the Android   ensuring that it considers the remote data (laptop ) as the valid one, ok?
----------------------------------------

*From: *Jiongxuan Zhang ***@***.***>
*To: *johannesjo/super-productivity ***@***.***>
*CC: *ElpyGh ***@***.***>; Mention ***@***.***>
*Date: *27 Jan 2025 15:49:26
*Subject: *Re: [johannesjo/super-productivity] Major Android Update Discussion: Introducing Offline Support and Resolving CORS Issues (Issue #3527)

**Comment by johannesjo** [comment]
> Ok, so because I use the Android app that doesn't provide the export/import feature, I should synchronize with a laptop, then remove the Android app, reinstall it and then synchronize again from the Android   ensuring that it considers the remote data (laptop ) as the valid one, ok?

Yes. You also should delete all data for your android app ideally.

**Comment by eclecticpassions** [comment]
> > Ok, so because I use the Android app that doesn't provide the export/import feature, I should synchronize with a laptop, then remove the Android app, reinstall it and then synchronize again from the Android   ensuring that it considers the remote data (laptop ) as the valid one, ok?
> 
> Yes. You also should delete all data for your android app ideally. 

I want to get the new offline version, so I unchecked sync (Dropbox), deleted cache, app data and forced stop, uninstalled it and downloaded the latest version of F-Droid. I opened it and surprised it showed all my info like nothing happened. How do I get it to start fresh? (I will resync it with my Dropbox). Where is all the data stored on Android so I can delete it? Thanks @johannesjo

**Comment by johannesjo** [comment]
You need to delete also the user data from where you deleted your cache.

**Comment by johannesjo** [comment]
@jiongxuan I think we can close this now, right?

**Comment by zroug** [comment]
For me, the sync feature on the Android app is still broken, which seems to be due to CORS. Sync works on the desktop app, but on both the web app and the Android app, I get `HttpNotOkAPIError: Hr`. This is expected for the web app because CORS isn't configured properly on the server, but it should work for the Android app.

**Comment by jiongxuan** [comment]
> [@jiongxuan](https://github.com/jiongxuan) I think we can close this now, right?

Okay, I agree

**Comment by jiongxuan** [comment]
> For me, the sync feature on the Android app is still broken, which seems to be due to CORS. Sync works on the desktop app, but on both the web app and the Android app, I get `HttpNotOkAPIError: Hr`. This is expected for the web app because CORS isn't configured properly on the server, but it should work for the Android app.

As @johannesjo said, You need to delete also the user data from where you deleted your cache and try again @zroug

**Comment by zroug** [comment]
You are right, it works now. 😃

Sorry, I was so determined to confirm that it was a CORS error, I don't know how I missed that. Thanks for helping me out.
